plugin/pkg/auth/authorizer/node/graph.go

```diff
@@ -0,0 +1,244 @@
+	PVNode
+	SecretNode
+)
+
+var nodeTypes = map[nodeType]string{
+	ConfigMapNode: "configmap",
+	NodeNode:      "node",
+	PodNode:       "pod",
+	PVCNode:       "pvc",
+	PVNode:        "pv",
+	SecretNode:    "secret",
+}
+
+// must be called under a write lock
+func (g *Graph) getOrCreateNode(nodeType nodeType, namespace, name string) *NamedNode {
+	if node, exists := g.getNode(nodeType, namespace, name); exists {
+		return node
+	}
+	return g.createNode(nodeType, namespace, name)
+}
+
+// must be called under a read lock
+func (g *Graph) getNode(nodeType nodeType, namespace, name string) (*NamedNode, bool) {
+	node, exists := g.nodes[nodeType][namespace][name]
+	return node, exists
+}
+
+// must be called under a write lock
+func (g *Graph) createNode(nodeType nodeType, namespace, name string) *NamedNode {
+	typedNodes, exists := g.nodes[nodeType]
+	if !exists {
+		typedNodes = map[string]map[string]*NamedNode{}
+		g.nodes[nodeType] = typedNodes
+	}
+
+	namespacedNodes, exists := typedNodes[namespace]
+	if !exists {
+		namespacedNodes = map[string]*NamedNode{}
+		typedNodes[namespace] = namespacedNodes
+	}
+
+	node := NewNode(nodeType, namespace, name, g.graph.NewNodeID())
+	namespacedNodes[name] = node
+	g.graph.AddNode(node)
+
+	return node
+}
+
+// must be called under write lock
+func (g *Graph) deleteNode(nodeType nodeType, namespace, name string) {
+	node, exists := g.getNode(nodeType, namespace, name)
+	if !exists {
+		return
+	}
+
+	// find existing neighbors
+	neighborsToRemove := []graph.Node{}
+	g.graph.VisitFrom(node, func(neighbor graph.Node) bool {
+		if g.graph.Degree(neighbor) == 1 {
+			neighborsToRemove = append(neighborsToRemove, neighbor)
+		}
+		return true
+	})
+	g.graph.VisitTo(node, func(neighbor graph.Node) bool {
+		if g.graph.Degree(neighbor) == 1 {
+			neighborsToRemove = append(neighborsToRemove, neighbor)
+		}
+		return true
+	})
+
+	// remove the node
+	g.graph.RemoveNode(node)
+	delete(g.nodes[nodeType][namespace], name)
+	if len(g.nodes[nodeType][namespace]) == 0 {
+		delete(g.nodes[nodeType], namespace)
+	}
+	if len(g.nodes[nodeType]) == 0 {
+		delete(g.nodes, nodeType)
+	}
+
+	// remove neighbors that are now edgeless
+	for _, neighbor := range neighborsToRemove {
+		g.graph.RemoveNode(neighbor)
+		n := neighbor.(*NamedNode)
+		delete(g.nodes[n.nodeType][n.namespace], n.name)
+		if len(g.nodes[n.nodeType][n.namespace]) == 0 {
+			delete(g.nodes[n.nodeType], n.namespace)
+		}
+		if len(g.nodes[n.nodeType]) == 0 {
+			delete(g.nodes, n.nodeType)
+		}
+	}
+}
+
+// AddPod sets up edges for the following relationships:
+//   secret -> pod
+//   configmap -> pod
+//   pvc -> pod
+//   pod -> node
+func (g *Graph) AddPod(pod *api.Pod) {
+	g.lock.Lock()
+	defer g.lock.Unlock()
+
+	g.deleteNode(PodNode, pod.Namespace, pod.Name)
+	podNode := g.getOrCreateNode(PodNode, pod.Namespace, pod.Name)
+	nodeNode := g.getOrCreateNode(NodeNode, "", pod.Spec.NodeName)
+	g.graph.SetEdge(NewNodeEdge(podNode, nodeNode, nodeNode))
```

Comment:
I'm assuming this is a guiding principle.  All edges point towards the KubernetesNodeVertex.  If so, doc the overall structure.